sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprArrayElementNode.java

```diff
@@ -0,0 +1,101 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS
+ * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
+ * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.oracle.truffle.llvm.runtime.debug.debugexpr.nodes;
+
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.InvalidArrayIndexException;
+import com.oracle.truffle.api.interop.UnknownIdentifierException;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.nodes.NodeInfo;
+import com.oracle.truffle.api.nodes.UnexpectedResultException;
+import com.oracle.truffle.llvm.runtime.debug.debugexpr.parser.DebugExprException;
+import com.oracle.truffle.llvm.runtime.debug.debugexpr.parser.DebugExprType;
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMExpressionNode;
+
+@NodeInfo(shortName = "[]")
+public class DebugExprArrayElementNode extends LLVMExpressionNode {
+
+    @Child private LLVMExpressionNode indexNode;
+    private DebugExprType type;
+    private Object member;
+    private Object baseMember;
+
+    public DebugExprArrayElementNode(Object baseMember, LLVMExpressionNode indexNode, DebugExprType type) {
+        this.indexNode = indexNode;
+        this.baseMember = baseMember;
+        this.type = type;
+    }
+
+    public DebugExprType getType() {
+        return type;
+    }
+
+    public Object getMember() {
+        return member;
+    }
+
+    @Override
+    public Object executeGeneric(VirtualFrame frame) {
+        int idx;
+        try {
+            idx = indexNode.executeI32(frame);
+        } catch (UnexpectedResultException e) {
+            try {
+                // in case of a complex expression as index (e.g. outerArray[innerArray[2]]), the
+                // index (=e.getResult()) is no Integer but a LLVMDebugObject$Primitive instead
+                idx = Integer.parseInt(e.getResult().toString());
+            } catch (NumberFormatException e1) {
+                throw DebugExprException.typeError(this, e.getResult());
+            }
+        }
+        InteropLibrary library = InteropLibrary.getFactory().getUncached();
+        if (library.hasMembers(baseMember)) {
+            Object getmembers;
+            try {
+                getmembers = library.getMembers(baseMember);
```

Comment:
@jkreindl is this really correct? Do our debug arrays really expose their array elements as members? Why not as array elements?

Reply:
I don't remember about the reason for that. Maybe the debugger didn't use to show array elements at some point (I was probably still using the NetBeans plugin then), or maybe I decided that I wanted a particular formatting for the array indices that matched lldb. We could change that though.

Reply:
Ok. No need to change it now, I was just wondering, because that code looks weird.